File: app/src/platform/localNode.ts
/**
 * Das eigene Gerät als Gegenstelle.
 *
 * <p>
 * **Eine Kopplung geht immer in beide Richtungen.** Wer koppelt, richtet damit
 * nicht nur eine Richtung ein: beide Seiten tauschen beim Koppeln aus, was sie
 * voneinander brauchen — in einem Aufruf, ohne zweiten Weg. Was danach zu tun
 * ist, erledigt jede Seite bei sich zu Hause, und dafür steht diese
 * Schnittstelle.
 * </p>
 *
 * <p>
 * **Der Vorgänger machte es andersherum**, und das war der Fehler: er reichte
 * einen Kopplungscode weiter, den die Gegenseite binnen fünf Minuten einlösen
 * musste. Damit hing die Gegenrichtung an einem laufenden Server, einem offenen
 * Fenster und einer Uhr — wer beim Koppeln die Freigabe nicht eingeschaltet
 * hatte, bekam sie nie.
 * </p>
 *
 * <p>
 * Alle vier Wege laufen nativ und nicht über HTTP: der eigene Agent weist sich
 * mit einem selbst ausgestellten Zertifikat aus, und die Seite müsste ihm erst
 * vertrauen, um ihn überhaupt fragen zu können. Am Handy führt der Weg über das
 * Plugin, im Fenster über die Brücke zur Wirtsanwendung.
 * </p>
 */
export interface LocalNode {
  /**
   * Ob es hier jemanden gibt, der die Liste der zugelassenen Geräte führt.
   *
   * <p>
   * **Nicht dasselbe wie ein vorhandener Steckbrief**, und die Verwechslung war
   * ein Fehler: die Gegenrichtung hing an {@link profile}, also an einer
   * Adresse. Ein Handy, das gerade in keinem Netz hängt, hat keine — und trug
   * den Schlüssel der Gegenseite deshalb **nicht** ein. Danach stand das Gerät
   * in der Liste und antwortete auf jede Anfrage mit „kenne ich nicht", ohne
   * dass irgendwo stand, warum. Eine Adresse ist eine Auskunft über das Netz von
   * jetzt; ein Eintrag in einer Datei gilt, sobald der Server startet.
   * </p>
   *
   * <p>
   * **Und warum es eine Frage bleibt und keine Eigenschaft:** am Handy führt die
   * Liste der Prozess selbst, hier ist die Antwort immer ja. Im Fenster führt sie
   * der Agent nebenan, und ob der läuft, ist eine Entscheidung des Nutzers — wer
   * nur andere Geräte steuern will, braucht ihn nicht. Ein „nein" ist dort keine
   * Störung, sondern der Normalfall, und darf deshalb keine Meldung auslösen.
   * </p>
   */
  ready(): Promise<boolean>

  /**
   * Der eigene Steckbrief — er geht mit, wenn dieses Gerät ein anderes koppelt.
   *
   * `undefined` heißt: dieses Gerät ist kein mögliches Ziel (kein Agent, keine
   * Adresse). Dann bleibt es bei der einen Richtung, und das ist kein Fehler.
   * Ob der Host gerade *läuft*, spielt ausdrücklich keine Rolle: der Steckbrief
   * beschreibt, wie dieses Gerät erreichbar wäre, und ein Eintrag in einer Datei
   * wirkt, sobald der Server startet.
   */
  profile(): Promise<DeviceProfile | undefined>

   /**
   * Die Steckbriefe, die beim Koppeln hier abgegeben wurden.
   *
   * Lesen leert den Eingang **nicht**. Es tat es einmal, und das war falsch:
   * ging danach irgendetwas schief, war der Steckbrief endgültig weg, und am
   * Bildschirm stand „noch kein Gerät gekoppelt" — ohne zweiten Versuch.
   */
  peers(): Promise<DeviceProfile[]>

  /**
   * Vergisst, was in der Liste steht. Erst jetzt: sonst käme ein Gerät, das
   * jemand aus seiner Liste entfernt hat, von allein zurück.
   */
  forget(ids: string[]): Promise<void>

  /**
   * Die Gegenrichtung eintragen: die Oberfläche der Gegenseite darf dieses
   * Gerät steuern. Ohne Code — der Schlüssel kam über eine Verbindung, an deren
   * Anfang genau ein Code stand.
   */
  grant(publicKey: string, label: string): Promise<void>

  /**
   * Der Ausweis dieses Geräts — das Schlüsselpaar, mit dem sich seine
   * Oberfläche bei fremden Geräten anmeldet.
   *
   * <p>
   * **Warum ihn die Gegenstelle führt und nicht die App:** bis zum 16.08.2026
   * lag er im Speicher der Weboberfläche, und die Gegenstelle kannte ihn nur,
   * weil die App ihn beim Start hinterlegte. Damit hing der Ausweis am
   * Lebenslauf einer Oberfläche — wer im Fenster nie die Fernsteuerung
   * anzeigte, hinterlegte nie etwas, und die Gegenseite bekam beim Koppeln ein
   * leeres `clientKey`. Jetzt liegt er dort, wo er hingehört: bei den übrigen
   * Schlüsseln des Geräts, und beide lesen dieselbe Stelle.
   * </p>
   *
   * <p>
   * `undefined` heißt: diese Umgebung führt keinen — im Browser ist das der
   * Normalfall. Dann legt die App selbst einen an und bewahrt ihn in ihrem
   * eigenen Schlüsselspeicher auf.
   * </p>
   */
  key(): Promise<ClientKey | undefined>
}

/**
 * Ein Schlüsselpaar, wie es die Gegenstelle herausgibt: ECDSA P-256, der
 * öffentliche Teil als Base64 im SPKI-Format, der private als Base64 im
 * PKCS-8-Format. Genau so nimmt es die WebCrypto-API des Browsers an.
 *
 * Schwesterfassungen: `setup/ClientKeyFile.cs` und `host/LocalClientKey.kt`.
 */
export interface ClientKey {
  publicKey: string
  privateKey: string
}

/** Was ein Gerät ist. Gegenstück zu `setup/DevicePlatform.cs`. */
export type DevicePlatform = 'windows' | 'android'

/**
 * Der Steckbrief eines Geräts: alles, was die Gegenseite braucht, um es später
 * von sich aus zu erreichen. Schwesterfassungen: `agent/Auth/DeviceProfile.cs`
 * und `host/DeviceProfile.kt`.
 */
export interface DeviceProfile {
  /**
   * Woran der Eingang diesen Eintrag wiedererkennt. Nur bei abgeholten
   * Steckbriefen gesetzt — der eigene braucht keine.
   */
  id?: string
  host: string
  port: number
  /** Wie das Gerät heißt. Für die Anzeige in der Geräteliste. */
  name: string
  /**
   * Womit es sich ausweist — `undefined` bei einem Zertifikat von Tailscale,
   * dem ohnehin jeder glaubt. Dann gibt es nichts zu bestätigen.
   */
  caFingerprint?: string
  /**
   * Der Fingerabdruck seines Agent-Schlüssels. Er ist die Kennung des Geräts in
   * der Liste: er bleibt gleich, auch wenn Name oder Adresse wechseln.
   */
  agentFingerprint?: string
  /**
   * Der öffentliche Schlüssel, mit dem sich die **Oberfläche** dieses Geräts
   * anmeldet. Er gehört in die `clients.json` der Gegenseite — das ist die
   * ganze Gegenrichtung, in einem Feld.
   */
  clientKey?: string
  /**
   * Ob dahinter ein Rechner oder ein Handy steckt — für das Symbol in der
   * Geräteliste. Er steht im Steckbrief und nicht nur in `/api/info`, damit
   * die Liste ihn auch dann zeigt, wenn das Gerät gerade aus ist.
   */
  platform?: DevicePlatform
}

/** Für Umgebungen, die selbst keine Gegenstelle sind — der Browser vor allem. */
export const noLocalNode: LocalNode = {
  ready: (): Promise<boolean> => Promise.resolve(false),
  profile: (): Promise<DeviceProfile | undefined> => Promise.resolve(undefined),
  peers: (): Promise<DeviceProfile[]> => Promise.resolve([]),
  forget: (): Promise<void> => Promise.resolve(),
  grant: (): Promise<void> => Promise.resolve(),
  key: (): Promise<ClientKey | undefined> => Promise.resolve(undefined),
}

/**
 * Was von einem Steckbrief übrig bleibt, wenn man ihn ernst nimmt.
 *
 * Dieselben Regeln wie auf beiden Agent-Seiten: eine Adresse, ein brauchbarer
 * Port. Unvollständiges wird verworfen statt halb benutzt — ein Steckbrief, an
 * dem etwas fehlt, führte später zu einem Fehlschlag an einer Stelle, an der
 * niemand mehr weiß, woher er kam.
 */
export function usableProfile(value: unknown): DeviceProfile | undefined {
  if (typeof value !== 'object' || value === null) {
    return undefined
  }

  const { id, host, port, name, caFingerprint, agentFingerprint, clientKey, platform } =
    value as Record<string, unknown>

  if (typeof host !== 'string' || host.trim().length === 0 || host.length > 255) {
    return undefined
  }

  if (typeof port !== 'number' || !Number.isInteger(port) || port < 1 || port > 65535) {
    return undefined
  }

  const address = host.trim()

  return {
    ...(typeof id === 'string' && id.length > 0 ? { id } : {}),
    host: address,
    port,
    // Ein leerer Eintrag in der Liste ließe sich später niemandem zuordnen.
    name: typeof name === 'string' && name.trim().length > 0 ? name.trim() : address,
    ...hex('caFingerprint', caFingerprint, 64),
    ...hex('agentFingerprint', agentFingerprint, 16),
    ...(typeof clientKey === 'string' && clientKey.trim().length > 0
      ? { clientKey: clientKey.trim() }
      : {}),
    // Ein unbekannter Wert zählt als keiner: die Liste zeigt dann kein Symbol,
    // und das ist besser als ein falsches.
    ...(platform === 'windows' || platform === 'android' ? { platform } : {}),
  }
}

function hex(field: string, value: unknown, length: number): Record<string, string> {
  if (typeof value !== 'string') {
    return {}
  }

  const trimmed = value.trim().toLowerCase()

  return trimmed.length === length && /^[0-9a-f]+$/.test(trimmed) ? { [field]: trimmed } : {}
}

File: desktop/LocalNode.cs
using System.Net.Http.Json;
using System.Text;
using System.Text.Json;

namespace RemoteDesktopClient;

/// <summary>
/// Dieser Rechner als Gegenstelle — gefragt wird der Agent nebenan, und wenn
/// der nicht läuft, sein Datenordner.
///
/// <para>
/// **Warum das Fenster fragt und nicht die Seite:** der Agent weist sich mit
/// einem selbst ausgestellten Zertifikat aus. Die Seite müsste ihm erst
/// vertrauen, um ihn überhaupt fragen zu können, wem man vertrauen soll — ein
/// Kreis, der sich nicht schließt. Hier gibt es ihn nicht: das Fenster läuft
/// auf demselben Rechner wie der Agent.
/// </para>
///
/// <para>
/// Deshalb wird das Zertifikat auf <c>127.0.0.1</c> auch nicht geprüft. Wer
/// dort lauscht, hat den Port 8443 dieses Rechners belegt — dann liefe der
/// eigene Agent gar nicht, und die Fernsteuerung wäre das kleinste Problem.
/// </para>
///
/// <para>
/// **Zwei Wege zu denselben Daten.** Läuft der Agent, geht alles über ihn: er
/// hält <c>clients.json</c> im Speicher, und eine Datei unter ihm zu ändern
/// ginge beim nächsten Schreiben verloren. Läuft er nicht, liest und schreibt
/// das Fenster die Dateien selbst — beim nächsten Start liest der Agent sie
/// ohnehin neu. Das ist der ganze Unterschied zwischen „koppeln geht nur mit
/// laufendem Agent" und „eingerichtet genügt": wer diesen Rechner nur zum
/// Steuern benutzt, soll den Agent nicht dafür starten müssen.
/// </para>
///
/// <para>
/// **Vier Wege statt zweier.** Bis Phase 31e reichte das Fenster einen
/// Kopplungscode weiter, den die Gegenseite binnen fünf Minuten einlösen
/// musste. Jetzt geht der Steckbrief bei der Kopplung mit, und was danach zu
/// tun ist, erledigt jede Seite bei sich zu Hause: den Schlüssel der anderen in
/// die eigene <c>clients.json</c>, ihren Steckbrief in die eigene Geräteliste.
/// </para>
/// </summary>
public static class LocalNode
{
    private static readonly HttpClient Client = Build();

    /// <summary>
    /// Der eigene Steckbrief: Adresse, Port, Name und die beiden Fingerabdrücke.
    ///
    /// Er hängt ausdrücklich **nicht** daran, ob dieser Rechner gerade steuerbar
    /// sein will — er beschreibt, wie er erreichbar wäre. Deshalb kommt er bei
    /// gestopptem Agent aus dem Datenordner statt gar nicht.
    /// </summary>
    /// <returns>
    /// <c>null</c>, wenn keine Adresse eingetragen ist. Dann ist dieser Rechner
    /// kein mögliches Ziel, und es bleibt bei einer Richtung.
    /// </returns>
    public static async Task<object?> SelfAsync(CancellationToken cancellationToken = default)
    {
        var running = await ReadAsync("/api/pair/self", "profile", cancellationToken);

        // Ein Agent, der antwortet, hat recht: er kennt sein eigenes Zertifikat
        // und nicht nur die Dateien, aus denen es entstanden ist. Antwortet er
        // mit `null`, ist das ebenfalls seine Auskunft — dann fehlt die Adresse,
        // und der Datenordner wüsste es auch nicht besser.
        if (running is { } answered)
        {
            return answered.ValueKind == JsonValueKind.Null ? null : answered;
        }

        return AgentData.Profile();
    }

    /// <summary>
    /// Die Steckbriefe, die beim Koppeln hier abgegeben wurden.
    ///
    /// Nur mit laufendem Agent, und das ist keine Einschränkung: dort landet
    /// überhaupt nur etwas, wenn ein anderes Gerät diesen Rechner gekoppelt
    /// hat — und dafür musste er antworten.
    /// </summary>
    public static async Task<object?> PeersAsync(CancellationToken cancellationToken = default) =>
        await ReadAsync("/api/pair/peers", "peers", cancellationToken);

    /// <summary>
    /// Vergisst die Steckbriefe, die in der Liste stehen. Erst nach dem
    /// Eintragen — vorher wäre ein Fehlschlag endgültig.
    /// </summary>
    public static Task ForgetAsync(
        string[] ids, CancellationToken cancellationToken = default) =>
        PostAsync("/api/pair/peers/forget", new { ids }, cancellationToken);

    /// <summary>
    /// Trägt die Oberfläche der Gegenseite in die <c>clients.json</c> dieses
    /// Rechners ein — die Gegenrichtung, ohne einen zweiten Aufruf über das Netz.
    ///
    /// Bei gestopptem Agent schreibt das Fenster die Datei selbst. Ein Eintrag
    /// wirkt, sobald der Agent startet; er hat keine Frist.
    /// </summary>
    public static async Task GrantAsync(
        string publicKey, string? label, CancellationToken cancellationToken = default)
    {
        if (await PostAsync(
                "/api/pair/grant",
                new { publicKey, label = label ?? string.Empty },
                cancellationToken))
        {
            return;
        }

        AgentData.Grant(publicKey, label ?? string.Empty);
    }

    /// <summary>
    /// Hebt die Gegenrichtung wieder auf: dieses Gerät darf diesen Rechner
    /// nicht mehr steuern.
    ///
    /// Bei laufendem Agent über ihn — nur so werden auch die Verbindungen
    /// getrennt, die gerade offen sind. Bei gestopptem Agent aus der Datei;
    /// offene Verbindungen kann es dann keine geben.
    /// </summary>
    public static async Task RevokeAsync(
        string clientId, CancellationToken cancellationToken = default)
    {
        if (await DeleteAsync($"/api/clients/{Uri.EscapeDataString(clientId)}", cancellationToken))
        {
            return;
        }

        AgentData.Revoke(clientId);
    }

    /// <summary>
    /// Ob der Agent gerade läuft.
    ///
    /// Am Rechner ist „dieses Gerät freigeben" keine Einstellung der
    /// Oberfläche, sondern eine Auskunft: freigegeben ist er, solange der Agent
    /// läuft. Ein Schalter dafür stünde in der App und meinte etwas, das ihr
    /// nicht gehört.
    /// </summary>
    public static async Task<bool> RunningAsync(CancellationToken cancellationToken = default)
    {
        try
        {
            using var response = await Client.GetAsync(
                $"https://127.0.0.1:{AgentData.AgentPort}/health", cancellationToken);

            return response.IsSuccessStatusCode;
        }
        catch (Exception ex) when (ex is HttpRequestException or TaskCanceledException)
        {
            return false;
        }
    }

    /// <summary>
    /// Ein frischer Kopplungscode.
    ///
    /// <para>
    /// Der eine Weg, der einen laufenden Agent wirklich voraussetzt: der Code
    /// lebt in seinem Speicher, und einlösen muss ihn ebenfalls er. Einen Code
    /// anzuzeigen, den niemand einlösen kann, wäre eine Einladung ins Leere —
    /// deshalb fliegt hier ein Fehler statt eines leeren Feldes.
    /// </para>
    /// </summary>
    public static async Task<JsonElement> CodeAsync(CancellationToken cancellationToken = default)
    {
        using var response = await Client.PostAsync(
            $"https://127.0.0.1:{AgentData.AgentPort}/api/pair/code",
            content: null,
            cancellationToken);

        response.EnsureSuccessStatusCode();

        return await response.Content.ReadFromJsonAsync<JsonElement>(cancellationToken);
    }

    /// <summary>
    /// Wer diesen Rechner steuern darf. Bei gestopptem Agent aus der Datei — die
    /// Liste gilt auch dann, sie wirkt nur gerade nicht.
    /// </summary>
    public static async Task<object?> ClientsAsync(CancellationToken cancellationToken = default)
    {
        var running = await ReadAsync("/api/clients", "clients", cancellationToken);

        return running is { } answered ? answered : AgentData.Clients();
    }

    /// <summary>
    /// Liest beim Agent und meldet, ob er überhaupt geantwortet hat.
    /// </summary>
    /// <returns>
    /// <c>null</c>, wenn kein Agent läuft oder er die Anfrage abgelehnt hat.
    /// Ein <c>JsonValueKind.Null</c> ist etwas anderes: dann hat er geantwortet,
    /// und die Antwort lautet „gibt es nicht".
    /// </returns>
    private static async Task<JsonElement?> ReadAsync(
        string path, string field, CancellationToken cancellationToken)
    {
        try
        {
            using var response = await Client.GetAsync(
                $"https://127.0.0.1:{AgentData.AgentPort}{path}", cancellationToken);

            if (!response.IsSuccessStatusCode)
            {
                return null;
            }

            var body = await response.Content.ReadFromJsonAsync<JsonElement>(cancellationToken);

            return body.TryGetProperty(field, out var value) ? value : null;
        }
        catch (Exception ex) when (ex is HttpRequestException or TaskCanceledException
                                       or JsonException)
        {
            // Kein Agent. Das ist der Normalfall auf einem Rechner, der nur
            // steuern und nicht gesteuert werden soll.
            return null;
        }
    }

    /// <summary>
    /// Schreibt beim Agent.
    /// </summary>
    /// <returns>
    /// <c>false</c>, wenn er nicht läuft — dann ist der Aufrufer dran, und zwar
    /// mit den Dateien. Ein Fehlschlag eines <em>laufenden</em> Agents fliegt
    /// dagegen weiter: die schreibenden Wege sind die Gegenrichtung selbst, und
    /// bleibt einer still liegen, sieht das später aus wie eine Kopplung, die
    /// nie angeboten wurde.
    /// </returns>
    private static async Task<bool> PostAsync(
        string path, object payload, CancellationToken cancellationToken)
    {
        using var content = new StringContent(
            JsonSerializer.Serialize(payload), Encoding.UTF8, "application/json");

        try
        {
            using var response = await Client.PostAsync(
                $"https://127.0.0.1:{AgentData.AgentPort}{path}", content, cancellationToken);

            response.EnsureSuccessStatusCode();

            return true;
        }
        catch (Exception ex) when (ex is HttpRequestException { StatusCode: null }
                                       or TaskCanceledException)
        {
            return false;
        }
    }

    /// <summary>
    /// Löscht beim Agent.
    /// </summary>
    /// <returns>
    /// <c>false</c>, wenn er nicht läuft — dann ist der Aufrufer dran. Ein
    /// 404 zählt als erledigt: was nicht da ist, muss nicht weg.
    /// </returns>
    private static async Task<bool> DeleteAsync(string path, CancellationToken cancellationToken)
    {
        try
        {
            using var response = await Client.DeleteAsync(
                $"https://127.0.0.1:{AgentData.AgentPort}{path}", cancellationToken);

            if (response.StatusCode == System.Net.HttpStatusCode.NotFound)
            {
                return true;
            }

            response.EnsureSuccessStatusCode();

            return true;
        }
        catch (Exception ex) when (ex is HttpRequestException { StatusCode: null }
                                       or TaskCanceledException)
        {
            return false;
        }
    }

    private static HttpClient Build()
    {
        var handler = new HttpClientHandler
        {
            ServerCertificateCustomValidationCallback = (_, _, _, _) => true
        };

        return new HttpClient(handler) { Timeout = TimeSpan.FromSeconds(4) };
    }
}
